lib/sprockets/redirect.rb

```diff
@@ -74,14 +76,36 @@ def digest_path
 
     # Sends a redirect header back to browser
     def redirect_to_digest_version(env)
-      url = URI(@asset_host || @request.url)
-      url.path = "#{@prefix}/#{digest_path}"
+      url = File.join(compute_asset_host, @prefix, digest_path)
 
       headers = { 'Location'      => url.to_s,
                   'Content-Type'  => Rack::Mime.mime_type(::File.extname(digest_path)),
                   'Pragma'        => 'no-cache',
                   'Cache-Control' => 'no-cache; max-age=0' }
-      [self.class.redirect_status, headers, [redirect_message(url.to_s)]]
+
+      [self.class.redirect_status, headers, [redirect_message(url)]]
+    end
+
+    def compute_asset_host
+      host = @asset_host
+
+      if host.respond_to?(:call)
+        arity = host.respond_to?(:arity) ? host.arity : host.method(:call).arity
```

Comment:
This logic is very complex.

I believe we don't need to check for arity. Just call with `logical_path, request`. User can ignore those in the lambda, or it would automatically get ignored if they use Proc.